zebedee-cms/src/main/java/com/github/onsdigital/zebedee/model/publishing/PublishNotification.java

```diff
@@ -1,98 +1,145 @@
 import com.github.onsdigital.zebedee.model.Collection;
+import com.github.onsdigital.zebedee.model.publishing.cacheAPI.NotificationPayloadCacheApi;
+import com.github.onsdigital.zebedee.model.publishing.cacheAPI.SendNotificationCacheApi;
 import com.github.onsdigital.zebedee.util.SlackNotification;
 import com.github.onsdigital.zebedee.util.slack.PostMessageField;
 import org.joda.time.DateTime;
 
 import java.text.SimpleDateFormat;
+import java.util.ArrayList;
 import java.util.Date;
 import java.util.List;
 
-import static com.github.onsdigital.logging.v2.event.SimpleEvent.info;
 import static com.github.onsdigital.logging.v2.event.SimpleEvent.error;
+import static com.github.onsdigital.logging.v2.event.SimpleEvent.info;
+import static com.github.onsdigital.zebedee.model.publishing.cacheAPI.NotificationPayloadCacheApi.BULLETIN_SEGMENT;
 
 /**
  * Created by bren on 16/12/15.
  * Publish notification to be sent to the website for caching timing purposes
  */
+
 public class PublishNotification {
 
+    private NotificationPayload payload;
+    private final List<NotificationPayloadCacheApi> payloads = new ArrayList<>();
+    private final String collectionId;
+
     private static final List<Host> websiteHosts;
     private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
-    private NotificationPayload payload;
 
     static {
         websiteHosts = Configuration.getWebsiteHosts();
     }
 
     public PublishNotification(Collection collection, List<String> urisToUpdate, List<ContentDetail> urisToDelete) {
+        this.collectionId = collection.getDescription().getId();
 
         // Delay the clearing of the cache after publish to minimise load on the server while publishing.
         Date clearCacheDate = new DateTime(collection.getDescription().getPublishDate())
                 .plusSeconds(Configuration.getSecondsToCacheAfterScheduledPublish()).toDate();
 
-        this.payload = new NotificationPayload(collection.getDescription().getId(), urisToUpdate, urisToDelete,
-                clearCacheDate);
+        if (Configuration.isLegacyCacheAPIEnabled()) {
+            if (urisToUpdate != null) {
+                for (String uriToUpdate : urisToUpdate) {
+                    NotificationPayloadCacheApi notificationPayloadCacheApi = new NotificationPayloadCacheApi(collection.getDescription().getId(), uriToUpdate, clearCacheDate);
+                    payloads.add(notificationPayloadCacheApi);
+                    addPayloadForBulletinLatest(notificationPayloadCacheApi.uriToUpdate, collection, clearCacheDate);
+                }
+            }
+        } else {
+            this.payload = new NotificationPayload(collection.getDescription().getId(), urisToUpdate, urisToDelete,
+                    clearCacheDate);
+        }
+
+    }
+
+    private void addPayloadForBulletinLatest(String uriToUpdate, Collection collection, Date clearCacheDate){
```

Comment:
All the changes you've made are a step in the right direction, however I think we can do more to make this code more clear and easy to test. For instance, it looks like you need 3 classes:
1. `LegacyCacheApiPayload`: this would be a very simple and dumb POJO with just a path, a release time and a collection ID. There is no logic in this class. It's just a DTO. So only getters and setters, and the `@SerializedName`s.
2. `LegacyCacheApiPayloadBuilder`: this could be a class that has only one public method, that accepts a collection ID, a list of paths, and a release time (and anything else you may need), and returns a collection of `LegacyCacheApiPayload`s. This is where you put all the logic, and where you can check that the list of paths is not null, that each path is not null, etc. This class always returns a list of **valid** payloads. If none of the inputs is valid (for instance, if you pass a list of paths, and all the paths are null), it returns an empty list. If some of the inputs are valid and some invalid, then it discards the invalid ones, and still returns a list of payloads. But the point is that all the validation and transformation is done here. And you can always be certain that the output of this function will always be a valid list of payloads. In this method, it's also when you would check for the Bulletin's `/latest` paths. So, for instance, one valid path that is a Bulletin `/latest` would generate two payload objects.
3. `LegacyCacheApiClient`: this is where you send the payloads to the API. It's essentially your `SendNotificationCacheApi`.

Does this make sense? What do you think?

Reply:
yes, I would think it would add value

Reply:
Thanks for implementing it. For the record, when I mentioned `LegacyCacheApiPayloadBuilder` I didn't mean to implement the builder pattern, apologies if that was confusing, I should have chosen a different name. That does not mean that using the builder pattern is wrong, I'm just mentioning this for clarity.